### PR TITLE
ddl: alter table add partition on hash partitioned table

### PR DESCRIPTION
**Pull request by ciscoxll**
### What problem does this PR solve? 
For https://github.com/pingcap/parser/pull/41

### What is changed and how it works?

Currently support alter table `add partition partitions` and `coalesce partition` by hash partitioned table syntax, because the hash type partitions add partition and coalesce partition involve more complicated mobile data, so we will implement the full function one after another. Here we will first report the error to the user.

### Check List 

Tests 

 - Unit test

---
This change is [Reviewable](https://reviewable.io/reviews/pingcap/tidb/8375)

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -4,6 +4,7 @@ require (
 	github.com/BurntSushi/toml v0.3.1
 	github.com/beorn7/perks v0.0.0-20180321164747-3a771d992973 // indirect
 	github.com/blacktear23/go-proxyprotocol v0.0.0-20171102103907-62e368e1c470
+	github.com/ciscoxll/parser v0.0.0-20181120113012-ce0923282643 // indirect
```
Comment (tiancaiamao) [inline go.mod]: It should be 

```
replace github.com/pingcap/parser => github.com/ciscoxll/parser v0.0.0-20181120113012-ce0923282643
```

at the end of the file.

Comment (ciscoxll) [reply go.mod]: @tiancaiamao Done

File: go.mod [hunk 2]
```diff
@@ -48,3 +48,5 @@ require (
 	google.golang.org/grpc v1.16.0
 	gopkg.in/natefinch/lumberjack.v2 v2.0.0
 )
+
+replace github.com/pingcap/parser => github.com/ciscoxll/parser v0.0.0-20181120113012-ce0923282643
```
Comment (tiancaiamao) [inline go.mod]: Parser is merged, remove here and update the reference

File: ddl/ddl_api.go [hunk 3]
```diff
@@ -1384,6 +1390,31 @@ func (d *ddl) AddTablePartitions(ctx sessionctx.Context, ident ast.Ident, spec *
...
+	meta := t.Meta()
+	if meta.GetPartitionInfo() == nil {
+		return errors.Trace(ErrPartitionMgmtOnNonpartitioned)
+	}
+
+	// We don't support coalesce partitions hash type partition now.
+	if t.Meta().Partition.Type == model.PartitionTypeHash {
+		return errors.Trace(ErrUnsupportedCoalescePartition)
+	}
+
```
Comment (tiancaiamao) [inline ddl/ddl_api.go]: What happen if a user coalesce partitions on a range partitioned table?
What if the coalesce partition Num = 0 ?

Comment (ciscoxll) [reply ddl/ddl_api.go]: @tiancaiamao Done.

**Comment by tiancaiamao** [comment]
LGTM

**Comment by ciscoxll** [comment]
@winkyao @crazycs520 PTAL

**Comment by ciscoxll** [comment]
@winkyao @crazycs520 PTAL

File: ddl/ddl_api.go [hunk 4]
```diff
@@ -1384,6 +1386,31 @@ func (d *ddl) AddTablePartitions(ctx sessionctx.Context, ident ast.Ident, spec *
...
+		return errors.Trace(infoschema.ErrTableNotExists.GenWithStackByArgs(ident.Schema, ident.Name))
+	}
+
+	meta := t.Meta()
+	if meta.GetPartitionInfo() == nil {
+		return errors.Trace(ErrPartitionMgmtOnNonpartitioned)
+	}
+
+	// We don't support coalesce partitions hash type partition now.
+	if t.Meta().Partition.Type == model.PartitionTypeHash {
```
[the remark this answers is not in the document]
Comment (ciscoxll) [reply ddl/ddl_api.go]: @crazycs520 Done.

**Comment by ciscoxll** [comment]
@winkyao @zimulala  PTAL

**Comment by ciscoxll** [comment]
@winkyao @zimulala PTAL

**Comment by ciscoxll** [comment]
@zimulala @winkyao PTAL

File: ddl/ddl_api.go [hunk 5]
```diff
@@ -1333,10 +1335,6 @@ func (d *ddl) AddColumn(ctx sessionctx.Context, ti ast.Ident, spec *ast.AlterTab
 
 // AddTablePartitions will add a new partition to the table.
 func (d *ddl) AddTablePartitions(ctx sessionctx.Context, ident ast.Ident, spec *ast.AlterTableSpec) error {
-	if len(spec.PartDefinitions) == 0 {
```
Comment (winkyao) [inline ddl/ddl_api.go]: Why you remove this？What if the Type == model.PartitionTypeRange and the len(spec.PartDefinitions) is 0?

File: ddl/ddl_api.go [hunk 6]
```diff
@@ -1351,6 +1349,16 @@ func (d *ddl) AddTablePartitions(ctx sessionctx.Context, ident ast.Ident, spec *
 	if meta.GetPartitionInfo() == nil {
 		return errors.Trace(ErrPartitionMgmtOnNonpartitioned)
 	}
+	// We don't support add hash type partition now.
+	if t.Meta().Partition.Type == model.PartitionTypeHash {
+		return errors.Trace(ErrUnsupportedAddPartition)
+	}
+
+	if t.Meta().Partition.Type == model.PartitionTypeRange {
```
Comment (winkyao) [inline ddl/ddl_api.go]: You can check `if len(spec.PartDefinitions) == 0` in buildPartitionInfo. and you can merge the two check together, like `if t.Meta().Partition.Type == model.PartitionTypeRange && if len(spec.PartDefinitions) == 0 `

File: ddl/ddl_api.go [hunk 7]
```diff
@@ -1384,6 +1390,36 @@ func (d *ddl) AddTablePartitions(ctx sessionctx.Context, ident ast.Ident, spec *
...
+		return errors.Trace(ErrPartitionMgmtOnNonpartitioned)
+	}
+
+	// Coalesce partition can only be used on hash/key partitions.
+	if t.Meta().Partition.Type == model.PartitionTypeRange {
+		return errors.Trace(ErrCoalesceOnlyOnHashPartition)
+	}
+
+	// We don't support coalesce partitions hash type partition now.
+	if t.Meta().Partition.Type == model.PartitionTypeHash {
```
[the remark this answers is not in the document]
Comment (ciscoxll) [reply ddl/ddl_api.go]: @winkyao Done.

**Comment by ciscoxll** [comment]
@winkyao PTAL.

**Comment by ciscoxll** [comment]
@winkyao PTAL

**Pull request merged**